botocore/endpoint_provider.py

```diff
@@ -377,6 +377,39 @@ def _not(self, value):
         """
         return not value
 
+    def aws_is_virtual_hostable_s3_bucket(self, value, allow_subdomains):
+        """Evaluates whether a value is a valid bucket name for virtual host
+        style bucket URLs. To pass, the value must meet the following criteria:
+        1. is_valid_host_label(value) is True
+        2. length between 3 and 63 characters (inclusive)
+        3. does not contain uppercase characters
+        4. is not formatted as an IP address
+
+        If allow_subdomains is True, split on `.` and validate
+        each component separately.
+
+        :type value: str
+        :type allow_subdomains: bool
+        :rtype: bool
+        """
+        if (
+            value is None
+            or value.lower() != value
+            or IPV4_RE.match(value) is not None
+            or (allow_subdomains is False and value.count(".") > 0)
+        ):
+            return False
+
+        if allow_subdomains is True:
+            return all(
+                self.aws_is_virtual_hostable_s3_bucket(label, False)
+                for label in value.split(".")
+            )
+
+        return 3 <= len(value) <= 63 and self.is_valid_host_label(
+            value, allow_subdomains=False
+        )
```

Comment:
We might be able to streamline this a bit more to rely on `is_valid_host_label`, but otherwise this looks great. Thanks!

```suggestion
        if (
            value is None
            or len(value) < 3
            or value.lower() != value
            or IPV4_RE.match(value) is not None
        ):
            return False

        if allow_subdomains is True:
            return all(
                self.aws_is_virtual_hostable_s3_bucket(label, False)
                for label in value.split(".")
            )

        return self.is_valid_host_label(value, allow_subdomains=False)
```